meta/src/main/scala/scalan/meta/ScalanParsers.scala

```diff
@@ -221,7 +222,16 @@ object ScalanAst {
     def isEntity(name: String) = entities.exists(e => e.name == name)
   }
 
-  def getConcreteClasses(defs: List[SBodyItem]) = defs.collect { case c: SClassDef => c }
+  //This function will filter out the concrete class
+  //implementations which inherit from the given trait.
+  def getConcreteClasses(entity : STraitDef, defs: List[SBodyItem]) : List[SClassDef] = {
+    //This function checks if the given entity is in the list of ansestors
+    def isAnAncestor(ancestors: List[STraitCall], entity : STraitDef) : Boolean = {
+      ancestors.find { case STraitCall(traitName ,_) => traitName == entity.name}.isDefined
+    }
+
+    defs.collect { case c : SClassDef => c}.filter{ c => isAnAncestor(c.ancestors, entity) }
```

Comment:
You can combine `collect` and `filter` here: `.collect { case c: SClassDef if isAnAncestor(c.ancestors, entity) => c}`. I'd also change `isAnAncestor` to take `c` instead of `ancestors`.